src/trie.py

```diff
@@ -23,11 +21,13 @@ def __init__(self, token=None, iterable=None):
     def insert(self, token):
         """
         Insert the value token into the trie.
-        <token> must be a string.
+        <token> must be a string and not contain '$' character.
         The token is ignored if already present in the trie.
         """
         if token and isinstance(type(token), type('word')):
             raise TypeError('the token must be a string')
+        if '$' in token:
```

Comment:
For situations like this, you definitely want to define a global constant as your default "end of token" char.

Or better yet, allow a keyword argument to `Trie.__init__` which sets the end-of-token char as an attribute of Trie instance.